### PR TITLE
Various enhancements made to pl-order-blocks

### PR DESCRIPTION
**Pull request by Frozemint**
**Main changes**:

This PR updates the element `pl-order-blocks` such that instead of using in-house styling, the element now uses Bootstrap. As you can see, the aesthetics of the element is vastly improved:

_Before_:

image

_After_:

image

Similar styling is applied to the submission and true answer fields, as shown: 

image

image

Possible areas of improvement is to thicken the text and borders slightly to improve contrast and visibility. Any feedback or suggestions are welcome. This PR is related to the #3476 Saga.
 
**Minor changes**:
This PR also updated the positioning of the bootstrap badges displaying the percentage score of the student submission for improved clarity and space saving. Invalid submissions are also now only displayed on 1 line for clarity and space saving.

image

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,8 +1,11641 @@
 {
     "name": "PrairieLearn",
     "version": "3.2.0",
-    "lockfileVersion": 1,
+    "lockfileVersion": 2,
     "requires": true,
+    "packages": {
```
Comment (SethPoulsen) [inline package-lock.json:6]: If there are no changes to the `package.json`, there shouldn't be any changes to the `package-lock.json`, right?

**Review by SethPoulsen — COMMENTED** [review]
The new UI does look really slick, but I do have some concerns. 

I think the new UI has a problem with discoverability and accessibility. Without the space in between the tiles, and the padding between the edge of the dropzone and the tiles, it is not clear at all that the tiles are actually draggable (or even, that they are tiles at all). This problem will be even worse for people who have imperfect eyesight (with respect to contrast or with respect to sharpness).

**Comment by SethPoulsen** [comment]
We also should make sure others who are planning on using this have a chance to comment if they wish. @mfsilva22 @zillesc @mfwlr @cheeren

File: package-lock.json (hunk 1 above)
Comment (Frozemint) [reply package-lock.json:6]: yeah, I dont know why `package-lock.json` keeps changing either

**Comment by Frozemint** [comment]
> The new UI does look really slick, but I do have some concerns.
> 
> I think the new UI has a problem with discoverability and accessibility. Without the space in between the tiles, and the padding between the edge of the dropzone and the tiles, it is not clear at all that the tiles are actually draggable (or even, that they are tiles at all). This problem will be even worse for people who have imperfect eyesight (with respect to contrast or with respect to sharpness).

Yes, I am concerned with the visuals as well, which is why I wanted to get some input from others. This is the best bootstrap element I could find, and even then the visibility is still somewhat lacking. I've tried modifying the visuals a little bit by thickening borders and increasing font weight, but doing so made the element look somewhat jarring in an uncanny valley way... since the rest of the bootstrap style are all usually hairline thin borders and bright whites....

File: package-lock.json (hunk 1 above)
Comment (SethPoulsen) [reply package-lock.json:6]: certain `npm` commands will change it as a way of keeping track of all the different versions of dependencies you have used in the past (perhaps among some other things, I'm not sure). If you haven't actually changed `package.json`, don't commit `package-lock.json`, keep it as it is. Now that you have committed it, you will need to revert it.

**Comment by SethPoulsen** [comment]
> doing so made the element look somewhat jarring in an uncanny valley way
hmm...what about leaving the border width and fonts but just putting the margins and shape of the tiles back how they were?

**Comment by cheeren** [comment]
I have not yet seen styling that I love for this type of problem. I agree that the Bootstrap option, as is, is too flat. What about using more white-space/padding and shadows? https://getbootstrap.com/docs/4.1/utilities/shadows/?  

You could also style it with "notches" (or something) so that each of the tiles has a little triangle cut out of the left edge, and the landing spot has a triangle on which to drop it (sort of like a jigsaw puzzle). I'm not sure how this would look/work with indentation.  I do, however, like the idea that a student could actually put a block in a specific position... seems like a more pedagogically sound way of building a solution incrementally. idk. i see problems with this, too.

Just to make sure you've seen it: https://runestone.academy/runestone/static/instructorguide/Directives/mixedUp.html  I don't think it's perfect... just a point of reference. it might be worth looking at the way scratch (block based programming language) handles its blocks.

i'm going to keep looking. bootstrap is so nice. i think my favorite idea right now is shadows, or some other 3d effect.

File: package-lock.json (hunk 1 above)
Comment (Frozemint) [reply package-lock.json]: reverted

**Comment by Frozemint** [comment]
_**I have reviewed and hereby sign the CLA**_

**Comment by SethPoulsen** [comment]
@Frozemint Are the changes here ready to be reviewed again? I don't have time now, but in few weeks I will have time to do a thorough review.

**Comment by Frozemint** [comment]
Yep, @SethPoulsen. This PR has gradually evolved from just a facelift, and some WIPs were also addressed as well. Please take as much time as you might need to look over everything.

File: elements/pl-fingerprint/info.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    "controller": "pl-fingerprint.py",
+    "dependencies": {
+        "elementStyles": [
+            "pl-fingerprint.css"
+        ],
+        "elementScripts": [
+            "pl-fingerprint.js"
+        ]
+    }
+}
```
Comment (SethPoulsen) [inline elements/pl-fingerprint/info.json:11]: Is all this `pl-fingerprint` stuff supposed to be in this PR?

File: elements/pl-order-blocks/pl-order-blocks.css [hunk 3]
```diff
@@ -30,73 +14,66 @@ div.pl-order-blocks-bottom {
...
-  max-height: none;
-  overflow: visible;
+li.ui-state-highlight { 
+  /*This styling applies to the placeholder answer tile that appears when the user is dragging */
+  height: 3em !important; 
+  border: none !important;
+  list-style-type: none;
+  background-color: #90a4ae !important;
+  min-width: 3em !important;
 }
```
Comment (SethPoulsen) [inline elements/pl-order-blocks/pl-order-blocks.css:30]: Do all these `!important`s need to be here? there are a lot of them in this PR and I'm not sure all (or any) of them are necessary.

File: elements/pl-order-blocks/pl-order-blocks.js [hunk 4]
```diff
@@ -2,9 +2,13 @@
 
 const TABWIDTH = 50;      // defines how many px the answer block is indented by, when the student
                           // drags and indents a block
-const INDENT_OFFSET = 10;  // For aesthetic, all answer blocks are offseted to the right by
+const INDENT_OFFSET = 0;  // For aesthetic, all answer blocks are offseted to the right by
                           // 5px, so the answer tiles are not directly touching the dropzone margins
```
Comment (SethPoulsen) [inline elements/pl-order-blocks/pl-order-blocks.js:6]: This comment seems to not match what the code is actually doing anymore. (as an aside, this is one of the reasons I dislike comments which simply restate what the code is doing)

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 5]
```diff
@@ -5,175 +5,167 @@
...
 PL_ANSWER_INDENT_DEFAULT = -1
 INDENTION_DEFAULT = False
+MAX_INDENTION_DEFAULT = 4
 SOURCE_BLOCKS_ORDER_DEFAULT = 'random'
 GRADING_METHOD_DEFAULT = 'ordered'
-MIN_INCORRECT_DEFAULT = None
-MAX_INCORRECT_DEFAULT = None
 SOURCE_HEADER_DEFAULT = 'Drag from here:'
-SOLUTION_HEADER_DEFAULT = 'Construct your solution here:'
+SOLUTION_HEADER_DEFAULT = 'Drop blocks here:'
```
Comment (SethPoulsen) [inline elements/pl-order-blocks/pl-order-blocks.py:17]: Why this change in wording?

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 6]
```diff
@@ -5,175 +5,167 @@
...
+
+        if check_indentation is False and answer_indent is not None:
+            raise Exception('<pl-answer> should not specify indentation if indentation is disabled.')
+
+        if is_correct:
+            correct_answers.append(answer_data_dict)
         else:
-            raise Exception('Tags nested inside <pl-order-blocks> must be <pl-answer>.')
+            incorrect_answers.append(answer_data_dict)
+        index += 1
```
Comment (SethPoulsen) [inline elements/pl-order-blocks/pl-order-blocks.py:87]: I always love to see code get shorter and simpler :)

**Comment by SethPoulsen** [comment]
Any idea why it looks completely different when I pull your code and run it on my machine? 

image

**Comment by SethPoulsen** [comment]
@Frozemint I would like to get these changes in as soon as we can now (in the next week or two), so that I can then push my Proof stuff on top of it (I know, I was going to do it months ago, but the semester got away from me...). I should be able to be pretty responsive on here now that the semester is over and I'm only focusing on research and development for the summer.

File: elements/pl-order-blocks/pl-order-blocks.js (hunk 4 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.js:6]: Yes, the offset was made obsolete after my refactoring (which was done for a code interview :laughing: ), this is why the comment wasn't updated... I'll be updating the code shortly

File: elements/pl-fingerprint/info.json (hunk 2 above)
Comment (Frozemint) [reply elements/pl-fingerprint/info.json:11]: oops, no... `pl-fingerprint` is something different that I worked on shortly for one of my profs... since it was quite urgent I didn't have time to branch properly and left the code in here :sweat_smile:, I'll be cleaning all the irrelevant code out shortly

**Comment by Frozemint** [comment]
> Any idea why it looks completely different when I pull your code and run it on my machine?
> 
> image

Sry Seth :sweat_smile:, it's supposed to look like that; I forgot to update the pictures in the post ever since I was working on under the hood changes...

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 6 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.py:87]: :sunglasses:

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 5 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.py:17]: I thought "construct your solution" was a little bit vague and unclear, whereas if we had the two headers say "Drag blocks from here" and "Drop blocks here", it will be more clear to the students. What do you think?

File: elements/pl-order-blocks/pl-order-blocks.css (hunk 3 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:30]: Yes. This is because the element uses bootstrap CSS as base, and we uses `!important` to override and adjust the bootstrap design CSS so the element wouldn't appear excessively "flat" and are more visible.

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 5 above)
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.py:17]: I like how the old version has the word 'solution' in it because it makes it clear that portion is what they are being graded on. But I'm not sure if it actually matters.

**Comment by SethPoulsen** [comment]
> > Any idea why it looks completely different when I pull your code and run it on my machine?
> 
> Sry Seth , it's supposed to look like that; I forgot to update the pictures in the post ever since I was working on under the hood changes...

I'm not really a fan of having the source and target block areas having the exact same color heading. I feel like they should have something visually different about them. I would be interested to hear what others think too.

**Comment by SethPoulsen** [comment]
Seems like a bug: 
image

**Review by SethPoulsen — COMMENTED** [review]
At this point I'm pretty happy with everything in this PR except 
1. the bug I showed where blocks can be indented in the wrong direction
2. various issues with the way things look: (1) answer key and 'your solution' look way different than the way the student input does and way different from one another, (2) everything is too big, and (3) same color on the source and target headings (I'd say this last one is not as big of a deal as the first two)

**Comment by SethPoulsen** [comment]
Looks like the python linter is failing on trailing whitespace. CLA assistant is failing but I don't know why because it looks like you did that right.

File: elements/pl-order-blocks/pl-order-blocks.py (hunk 5 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.py:17]: Yeah, especially since we have implemented a help button... I guess we shouldn't fix what isn't broken? I haven't heard any student feedback since the `pl-order-blocks` element went live tho.

File: elements/pl-order-blocks/pl-order-blocks.css [hunk 7]
```diff
@@ -30,73 +14,66 @@ div.pl-order-blocks-bottom {
...
-  border: none !important;
-  color: #efefff !important; 
-  background-color: #efefff !important; 
+ul.ui-sortable li {
+  margin-bottom: 5px;
+  border: 1px solid grey !important;
+  border-radius: 5px !important;
+  background-color: #f5f5f5;
+  font-size: 1.1em;
+  line-height: 1.1em;
```
[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: The line height is added because without it, the blocks looks not tall enough, and the font-size is increased slightly to slightly improve readability. I guess there might be a better way to do this?

File: elements/pl-order-blocks/pl-order-blocks.css [hunk 8]
```diff
@@ -30,73 +14,66 @@ div.pl-order-blocks-bottom {
...
-code.user-output {
-  border: 2px solid #A9A9A9; /* High visibility borders for accessibility */
-  margin-right: 4px;
+ul.answer_table li {
+  margin-bottom: 7px;
+  border: 2px solid grey !important;
+  border-radius: 5px !important;
+  background-color: #f5f5f5;
+  font-size: 1.2em;
+  line-height: 1.1em;
```
[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:64]: Same as previous

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 9]
```diff
@@ -254,58 +252,33 @@ def render(element_html, data):
...
-        student_answer_ranking = []
-        pl_drag_drop_element = lxml.html.fragment_fromstring(element_html)
+    if grading_mode == 'ranking':
+        index = 0
         for answer in student_answer:
-            e = pl_drag_drop_element.xpath(f'.//pl-answer[text()="{answer}"]')
-            is_correct = pl.get_boolean_attrib(e[0], 'correct', PL_ANSWER_CORRECT_DEFAULT)  # default correctness to True
-            if is_correct:
-                ranking = pl.get_integer_attrib(e[0], 'ranking', 0)
+            search = next((item for item in correct_answers if item['inner_html'] == answer['inner_html']), None)
```
[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.py:275]: only took a complete reimagination of the backend code :sunglasses:

**Comment by Frozemint** [comment]
> Why does this part:

It's a bug. I haven't paid too much attention to the styling of the submission and answer panel, since the bulk of the work was focused on making sure the question blocks worked. This should be an easy fix to copy the style to the submission and answer panel.

File: elements/pl-order-blocks/pl-order-blocks.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: > the blocks looks not tall enough
I don't think so, I think they perfect without the size increase. Have you had students complain about the size? Or are there certain types of content that don't fit? 

I really think we shouldn't mess with the size here, just leave it default.

File: elements/pl-order-blocks/pl-order-blocks.py [hunk 10]
```diff
@@ -5,175 +5,167 @@
...
-    else:
-        return 'badge-warning'
+def filter_multiple_from_array(data, keys):
+    result = []
+    for item in data:
+        dic = {}
+        for key in keys:
+            dic[key] = item[key]
+        result.append(dic)
+    return result
```
[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.py:33]: Yes, but I didnt figure out the one-liner way to do it when I started refactoring :laughing:

File: elements/pl-order-blocks/pl-order-blocks.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: Nah, I just liked how it looked with the increased line height :laughing:, I'll leave it default then, since I also saw how they looked without size increase when debugging CSS and they also look OK

[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: Wait, I found that if you disable the line-height, the element actually becomes bigger :neutral_face:, this might be a bootstrap thing

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: just change it to 1em then I guess? That's what I did locally and it was what I actually expected it to look like.

[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: image
When there are only one line for the block, `line-height: 1.1em` definitely seems overkill, but when you have multi-line text, it's barely adequate as the text starts to seem bunched

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: did you also change the font-size back to 1em?

[the remark this answers is not in the document]
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: Yes

[the remark this answers is not in the document]
Comment (SethPoulsen) [reply elements/pl-order-blocks/pl-order-blocks.css:55]: This is very odd, it look like this for me, which I don't see any issue with:
image

**Comment by SethPoulsen** [comment]
> Here is the fixed submission and answer panels

This is much much better. Maybe if we get rid of the shadow on the 'correct answer' and 'submitted' answer box it will make it clear that they aren't draggable? (in contrast with the input boxes). I also don't think we need to make the boxes all green, that looks a little funny.

Also, it seems that the 'correct answer' panel isn't showing the indentation at all as it should.

**Comment by Frozemint** [comment]
> Maybe if we get rid of the shadow on the 'correct answer' and 'submitted' answer box 
I could get rid of the shadow, but I think the blocks would look too flat and even out of place. As for the correct answer also showing the correct indent, that would take a bit more work. I'll get to that soon if it's needed

**Comment by SethPoulsen** [comment]
Lets not worry about the indentation thing then, it can come in a later PR. It should be on the todo list though probably. 

How did you have the CSS set up for this screenshot (from the top of this page)?

image

I like the way this looks.

**Comment by Frozemint** [comment]
> I like the way this looks.

I could revert the styling to that then?

**Comment by SethPoulsen** [comment]
> > I like the way this looks.
> 
> I could revert the styling to that then?

Yeah I think that would be good, but make sure the "Your answer" and "correct answer" boxes are the same width. Sorry for not being more clear earlier. 

After this change, I think we can have @mwest1066 have a look at the PR and consider merging it in.

File: elements/pl-order-blocks/pl-order-blocks.css [hunk 11]
```diff
@@ -30,73 +14,59 @@ div.pl-order-blocks-bottom {
...
-  height: 2em !important; 
-  line-height: 1.5em !important;
-  border: none !important;
-  color: #efefff !important; 
-  background-color: #efefff !important; 
+ul.ui-sortable li, ul.answer_table li, ul.submission_table li {
+  margin-bottom: 5px;
+  border: 1px solid grey !important;
+  border-radius: 5px !important;
+  line-height: 1.1em;
```
Comment (SethPoulsen) [inline elements/pl-order-blocks/pl-order-blocks.css:53]: I did a little more playing around with the font-size/line-height issue. If you just change this to 1.2em, it helps with the text not being too crowded in multi-line blocks, but without having the blocks get too big (as they were before when line-height and font-size were higher).

**Comment by Frozemint** [comment]
> Yeah I think that would be good, but make sure the "Your answer" and "correct answer" boxes are the same width. Sorry for not being more clear earlier.

image
Does this look good? The two boxes do not have same width, but that's due to my using `display: inline-block` to set the width to equal the content width.

**Comment by SethPoulsen** [comment]
Looks good to me.

Lets see if @mwest1066  can have a look at merging this.

File: elements/pl-order-blocks/pl-order-blocks.css (hunk 11 above)
Comment (Frozemint) [reply elements/pl-order-blocks/pl-order-blocks.css]: Yep, implemented your suggestion

**Comment by Frozemint** [comment]
Yeet

**Review by SethPoulsen — APPROVED** [review]

**Comment by SethPoulsen** [comment]
@Frozemint you need to copy paste the CLA agreement text _without_ the formatting for it to work (see over here: https://github.com/PrairieLearn/PrairieLearn/pull/4259)

**Comment by Frozemint** [comment]
I have reviewed and hereby sign the CLA

**Review by mwest1066 — APPROVED** [review]

**Comment by SethPoulsen** [comment]
@Frozemint I thought you had fixed this backwards indent bug but it appears to still be a thing: 
image

Whats the easiest way to fix this? a line of javascript somewhere?

**Comment by Frozemint** [comment]
> Whats the easiest way to fix this? a line of javascript somewhere?
I never knew this bug existed 😐 😐 😐 , maybe you could fix it by forcing the indent to be non-negative in the javascript? that's my best guess 😅
